libs/auth/src/cast_vote_record_hashes.ts

```diff
@@ -0,0 +1,308 @@
+    cvrIdLevel2Prefix: string;
+    cvrId: string;
+    cvrHash: string;
+  }
+): void {
+  client.run(
+    `
+    insert or replace into cvr_hashes (
+      cvr_id_level_1_prefix,
+      cvr_id_level_2_prefix,
+      cvr_id,
+      cvr_hash
+    ) values (?, ?, ?, ?)
+    `,
+    cvrIdLevel1Prefix,
+    cvrIdLevel2Prefix,
+    cvrId,
+    cvrHash
+  );
+}
+
+/**
+ * Gets the cast vote record root hash.
+ *
+ * See {@link CAST_VOTE_RECORD_HASHES_TABLE_SCHEMA} for more context.
+ */
+export function getCastVoteRecordRootHash(client: Client): string {
+  const row = client.one(
+    `
+    select cvr_hash as cvrHash
+    from cvr_hashes
+    where
+      cvr_id_level_1_prefix = '${NULL_VALUE}' and
+      cvr_id_level_2_prefix = '${NULL_VALUE}' and
+      cvr_id = '${NULL_VALUE}'
+    `
+  ) as { cvrHash: string } | undefined;
+  return row?.cvrHash ?? '';
+}
+
+/**
+ * Inserts a cast vote record hash (a leaf node in the Merkle tree) and updates the cast vote
+ * record root hash (the root node in the Merkle tree), updating relevant intermediate hashes along
+ * the way (intermediate nodes in the Merkle tree).
+ *
+ * See {@link CAST_VOTE_RECORD_HASHES_TABLE_SCHEMA} for more context.
+ */
+export function updateCastVoteRecordHashes(
+  client: Client,
+  cvrId: string,
+  cvrHash: string
+): void {
+  const cvrIdLevel1Prefix = cvrId.slice(0, 2);
+  const cvrIdLevel2Prefix = cvrId.slice(0, 4);
+
+  client.transaction(() => {
+    insertCastVoteRecordHash(client, {
+      cvrIdLevel1Prefix,
+      cvrIdLevel2Prefix,
+      cvrId,
+      cvrHash,
+    });
+
+    const siblingLeafHashes = selectCastVoteRecordHashes(client, {
+      cvrIdLevel1PrefixConstraint: { type: '=', value: cvrIdLevel1Prefix },
+      cvrIdLevel2PrefixConstraint: { type: '=', value: cvrIdLevel2Prefix },
+      cvrIdConstraint: { type: '!=', value: NULL_VALUE },
+      orderBy: 'cvr_id',
+    });
+    const level2Hash = sha256(siblingLeafHashes.join(''));
+    insertCastVoteRecordHash(client, {
+      cvrIdLevel1Prefix,
+      cvrIdLevel2Prefix,
+      cvrId: NULL_VALUE,
+      cvrHash: level2Hash,
+    });
+
+    const siblingLevel2Hashes = selectCastVoteRecordHashes(client, {
+      cvrIdLevel1PrefixConstraint: { type: '=', value: cvrIdLevel1Prefix },
+      cvrIdLevel2PrefixConstraint: { type: '!=', value: NULL_VALUE },
+      cvrIdConstraint: { type: '=', value: NULL_VALUE },
+      orderBy: 'cvr_id_level_2_prefix',
+    });
+    const level1Hash = sha256(siblingLevel2Hashes.join(''));
+    insertCastVoteRecordHash(client, {
+      cvrIdLevel1Prefix,
+      cvrIdLevel2Prefix: NULL_VALUE,
+      cvrId: NULL_VALUE,
+      cvrHash: level1Hash,
+    });
+
+    const level1Hashes = selectCastVoteRecordHashes(client, {
+      cvrIdLevel1PrefixConstraint: { type: '!=', value: NULL_VALUE },
+      cvrIdLevel2PrefixConstraint: { type: '=', value: NULL_VALUE },
+      cvrIdConstraint: { type: '=', value: NULL_VALUE },
+      orderBy: 'cvr_id_level_1_prefix',
+    });
```

Comment:
What do you think of the naming:
`siblingLeafHashes` -> `cvrHashesWithLevel2Prefix`
`siblingLevel2Hashes` -> `level2HashesWithLevel1Prefix`
`level1Hashes1` <-> `level1Hashes`

Reply:
Up to you, maybe this makes less sense. But I often find `sibling` terminology confusing because it's unclear if it will include the original node, but in these cases it does.

Reply:
Good point about the ambiguity of sibling and whether or not it includes the original! My only qualm with `cvrHashesWithLevel2Prefix` is that the hashes don't have the prefix, the IDs do. Easily solved though, with something like `cvrHashesForLevel2Prefix`. Will think some more and update

Reply:
https://github.com/votingworks/vxsuite/pull/3898/commits/26c3d1adb5b492dce2108b245cb24dd33e314f53